---
title: "lymphedema Transcripts Per Kilobase Million TPMs per Ensembl gene ID"
author: "Janis Corona"
date: "12/1/2020"
output: html_document
---


This study GSE132936 examines 10 patients who have malignant adipose tissue or excessive body fat that causes lymphedema in the limbs. These patients had liposuction from their upper abdominals where no diseased fat cells were located and also in their upper thighs where the diseased fat cells were located. The mesenchymal stem cells or MSCs were examined for their immune properties by looking at transcript counts or transcripts per kilobase Million TPM of each ENSEMBL gene ID. The cytokines and immunosuppression properties were targeted to see how the stem cells of the diseased fat (thighs) compare to the non-diseased fat (abs). A helpful article on understanding reads per kilobase million and transcripts per kilobase million can be found with an internet search for RPMs or TPMs. I found a helpful article, 'RPKM, FPKM and TPM, clearly explained | RNA-Seq Blog (rna-seqblog.com)' at
https://rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/


My personal summary of mesenchymal stem cells, they are cells that can be stemmed from original cells of fat, cartilage, blood, and muscle tissue as well as heart tissue and other organ tissues. They don't have the immune markers that compromise a person's immune system in healing from damage due to a wound or injury or disease when influenced by antibodies on the diseased cell version of the stem cells. For more information on stem cells or a better summary or understanding, there are a number of helpful topics to search online, and I base this summary of stem cells on the information made available at https://www.sciencedirect.com/topics/neuroscience/mesenchymal-stem-cell.


This GEO database, GSE132936, from NCBI is relatively new as it was published last June 2019 and can be accessed at: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132936

Transcripts are the ribonucleic transcripts in messenger RNA that is transcribed into the gene with possible variations or copy number variants (cnv) of each gene. These play functions with how the body transciption process is affected by epigenetic and hormonal or diet or stress regulators that can modify the transcription process. In this case, their is adipose stem cells being compared in how they are being transcribed or built into a normal abdominal fat cell compared to the diseased upper thigh fat cell. 

The original stem cells for these adipose cells are compared to the changes in transcription for the proportion of each gene in the diseased compared to normal adipose cell. When using the fold change or differential expression the diseased/healthy and the diseased-healthy are used respectively to target genes by their ENSEMBL ID as gene targets for therapeutic treatment of patients with the malignant adipose cells. Liposuction is a therapeutic treatment for these patients to keep the diseased fat (called lip edema as in lipid). This disease, lip edema, has 5 stages that start normal in stage 1 and not so obvious in stage 2, but starts getting progressively worse with each stage up to stage 5 that makes the patient look morbidly obese. Some people don't know they have it until they see that they are otherwise healthy, but follow a decent diet and workout without seeing any noticeable results.

These malignant adipose cells form largely on top of other clusters of the same cells creating rolls and folds that block lymphatic vessels and lymph nodes, preventing them from eliminating their share of tissue waste and fluids when working alongside the veinous system. This causes a back up of fluids in those tissues affected by diseased adipose cells and thus lymphedema that can be removed with elevation of limbs above the heart to drain, wearing compression garments to push the fluid out of the tissue in the day, working out, and also manual lymphatic drainage massage. 

I manually used Excel to gather each of the 20 samples, place in a separate excel worksheet for each sample, compare the metadata header info for each sample, and rename each sample of the GSM...40-GSM...59 with patient1-patient10 and the appended abs or thighs. With abs being normal or healthy fat cells and thighs being the diseased fat cells. All files can be found at my github repository, https://github.com/JanJanJan2018/lymphedema-stem-cells-GSE132936-for-lipedema, if they fit or can be compressed to a size less than 25 mb in size. You could also get the direct data at the link above with access ID GSE132936 and grab each tab delimited file for each of the 20 samples. The worksheet ended up being 179 MB and compressed to 177 MB, so it unfortunately couldn't be added. But all the orginal 20 files are in this repository.
***

Lets read in the data.
```{r}
TPMs <- read.csv('GSE132936_TPMs.csv', sep=',', header=T, 
                 na.strings=c('',' ','NA'),stringsAsFactors = F)
```

```{r}
head(TPMs,10)
```


```{r}
colnames(TPMs)
```

```{r}
healthy <- grep('abs',colnames(TPMs))
diseased <- grep('thighs',colnames(TPMs))

Healthy <- TPMs[,c(1,healthy)]
Diseased <- TPMs[,c(1,diseased)]
```


```{r}
Healthy$HealthyMean <- apply(Healthy[2:10],1,mean)
head(Healthy)
```

```{r}
Diseased$DiseasedMean <- apply(Diseased[2:11],1,mean)
head(Diseased)
```


```{r}
Stats <- cbind(Healthy,Diseased[2:12])
Stats$FoldChange <- Stats$DiseasedMean/Stats$HealthyMean
Stats$DifferentialExpression <- Stats$DiseasedMean-Stats$HealthyMean
```

```{r}
colnames(Stats)
```


```{r}
colnames(Stats) <- gsub('abs','healthy',colnames(Stats))
colnames(Stats) <- gsub('thighs','diseased',colnames(Stats))
colnames(Stats)
```

```{r}
head(Stats)
```

```{r}
range(Stats$FoldChange)
range(Stats$DifferentialExpression)
```

The fold change values of the Stats table need some modifications due to the errors in division by zero for the means. There are zeros, Inf, and NaNs in the fold change values. The zeros are when the diseased values are zero, the Inf's are from division by zero, and the NaNs are for the zeros divided by zero. Lets make this a conditional fold change feature.
```{r}
Stats1 <- round(Stats[,2:25],2)
Stats1$Ensembl_ID <- Stats$Ensembl_ID
Stats1 <- Stats1[,c(25,1:24)]
Stats1 <- Stats1[,c(1:11,13:22,12,23:25)]
colnames(Stats1)
```


```{r}
range(Stats1$FoldChange)
```

For the purposes of fold change, the diseased samples that have values and the healthy that don't, skew the data greatly, and indecisive on how to approach those values, as they make foldchange values astronomical when a diseased value is 39 as an examples and a healthy value is 0 it could have a positive 9 digit value for fold change. So we will subset and exclude the Inf, NaN, and 0 values for fold change
```{r}
StatsValidFCs <- subset(Stats1, Stats1$HealthyMean!=0)
range(StatsValidFCs$FoldChange)
```

```{r}
write.csv(StatsValidFCs,'lymphedemaStemCells_stats.csv',row.names=F)
```

```{r}
statsOrdered <- StatsValidFCs[order(StatsValidFCs$FoldChange,StatsValidFCs$DifferentialExpression,decreasing=T),]

FCsDoubled <- StatsValidFCs[StatsValidFCs$FoldChange>2,]
FCsHalved <- StatsValidFCs[StatsValidFCs$FoldChange<.5,]

quantile(FCsDoubled$FoldChange,c(.25,.5,.75,1))
quantile(FCsHalved$FoldChange,c(.25,.5,.75,1))

FCsDoubled2 <- FCsDoubled[FCsDoubled$FoldChange>4,]
FCsHalved2 <- FCsHalved[FCsHalved$FoldChange>.36,]


```

These are the top 75th percentile of genes transcripts with fold change values filtered by being at least doubled or halved in disease/healthy ratio. Where the top 25% of instances from each separate data set is used. This reduces the target genes to filter quite a bit.


```{r}
write.csv(FCsDoubled2,'transcripts75thPercentileDoubled.csv',row.names=F)
write.csv(FCsHalved2,'transcripts75thPercentileHalved.csv',row.names=F)
```


This study looked at two cytokines or immune genes, CDK1 and CHI3L1. We only have the ENSEMBL IDs but going to genecards.org should give us the ENSEMBL IDs to compare. CDK1 is ENSG00000170312 and CHI3L1 is ENSG00000133048. Lets look at these two genes in our data table.
```{r}
CDK1 <- grep('ENSG00000170312',Stats$Ensembl_ID)
CHI3L1 <- grep('ENSG00000133048',Stats$Ensembl_ID)

CDK1;CHI3L1
```

This data table of ENSEMBLE IDs doesn't have the mentioned genes by gene symbol. Even though these look like ENSEMBL IDs they are not. The platform to this series wasn't available. When entering each ID in this series into genecards.org it shows the gene symbol and a different ENSEMBL ID. For example, this data table's ID, ENST00000420783, is WDR6 for the gene symbol, but the ENSEMBL ID is ENSG00000178252. This is because we are looking at transcript data and not gene data directly. Hence, 'ENST' compared to 'ENSG' for transcript and gene respectively.

However, scrolling through the other data in genecards.org for CDK1, the transcripts are made available:
Transcripts for CDK1 Gene
Products: CRISPR  miRNA  Inhibitory RNA  Clone
mRNA/cDNA for CDK1 Gene 
6 REFSEQ mRNAs :
NM_001170406.1 NM_001170407.1 NM_001320918.1 NM_001786.5 NM_033379.4 See All 6 »
16 NCBI additional mRNA sequence :
AA459484.1 AK291939.1 AK295741.1 AW976143.1 BC014563.1 See All 16 »
9 Ensembl transcripts including schematic representations, and UCSC links to gene/alias where relevant :
ENST00000519078 723 nts (view in UCSC) ENST00000475504 1720 nts (view in UCSC) ENST00000395284 1889 nts (view in UCSC) ENST00000519760 276 nts (view in UCSC) ENST00000448257 1948 nts (view in UCSC) ENST00000373809 1613 nts (view in UCSC)ENST00000487784 598 nts (view in UCSC)ENST00000316629 1733 nts (view in UCSC)ENST00000614696 1884 nts (view in UCSC)

And for CHI3L1:
Transcripts for CHI3L1 Gene
Products: CRISPR  miRNA  Inhibitory RNA  Clone
mRNA/cDNA for CHI3L1 Gene 
1 REFSEQ mRNAs :
NM_001276.4
15 NCBI additional mRNA sequence :
AB209459.1 AK095458.1 AK130142.1 AK130200.1 AK225266.1 See All 15 »
5 Ensembl transcripts including schematic representations, and UCSC links to gene/alias where relevant :
ENST00000473185 1920 nts (view in UCSC) ENST00000255409 1747 nts (view in UCSC) ENST00000404436 921 nts (view in UCSC) ENST00000478742 825 nts (view in UCSC) ENST00000472064 890 nts (view in UCSC)

```{r}
CDK1_list <- c('ENST00000519078','ENST00000475504','ENST00000395284','ENST00000519760','ENST00000448257','ENST00000373809','ENST00000487784','ENST00000316629','ENST00000614696')

CHI3L1_list <- c('ENST00000473185','ENST00000255409','ENST00000404436','ENST00000478742','ENST00000472064')

```


```{r}
library(dplyr);library(tidyr)
```

```{r}
cdk1 <- Stats[Stats$Ensembl_ID %in% CDK1_list,]
chi3l1 <- Stats[Stats$Ensembl_ID %in% CHI3L1_list,]

cdk1$transcriptGene <- 'CDK1'
chi3l1$transcriptGene <- 'CHI3L1'

cdk1_chi3l1 <- rbind(cdk1,chi3l1)
cdk1_chi3l1
```

Only 8/9 transcripts were available for CDK1 and all 5/5 transcripts for CHI3L1 were available. We can see from the table of transcripts for these two genes, that most of the gene transcripts, if any data available, were more than two fold in transcription in the diseased versus the healthy adipose cells. Two transcripts of the CDK1 gene had no data (ENST00000519760) or were negligible for (ENST00000373809).

We can write this table out to csv and choose to examine later or map out.
```{r}
write.csv(cdk1_chi3l1,'cdk1_chi3l1.csv',row.names=F)
```

The study hasn't been written, but the summary of this data says that the inhibition of CDK1 and CHI3L1 gene transcripts could keep the proliferation and adipogenic properties of diseased adipose stem cells (ASC) from occuring. As these two cytokines seem to have a pivotal role in disease onset based on the up regulation of their transcripts in the diseased cells compared to normal phenotype ASC.


```{r}
source('geneCards2.R')
```

```{r}
Gene_Path <- './transcripts'

if (dir.exists(Gene_Path)){
  unlink(Gene_Path, recursive=TRUE)
  dir.create(Gene_Path)
} else {
  dir.create(Gene_Path)
}

for (i in FCsDoubled2$Ensembl_ID[1:2978]){
  transcript2Gene(i)
}



```

```{r}
doubledGenes <- read.csv('transcripts/geneTranscriptTable.csv',header=F,sep=',')
headerDoubled <- read.csv('transcripts/transcriptHeader.csv',header=T)
colnames(doubledGenes) <- headerDoubled$x

```


```{r}
Gene_Path <- './transcripts'

if (dir.exists(Gene_Path)){
  unlink(Gene_Path, recursive=TRUE)
  dir.create(Gene_Path)
} else {
  dir.create(Gene_Path)
}
```


```{r}
for (i in FCsDoubled2$Ensembl_ID[2979:length(FCsDoubled2$Ensembl_ID)]){
  transcript2Gene(i)
}



```


```{r}
doubledGenes2 <- read.csv('transcripts/geneTranscriptTable.csv',header=F,sep=',')
colnames(doubledGenes2) <- headerDoubled$x

```



```{r}
for (i in FCsHalved2$Ensembl_ID){
  transcript2Gene(i)
}




```


```{r}
halvedGenes <- read.csv('transcripts/geneTranscriptTable.csv',header=F,sep=',')
headerHalved <- read.csv('transcripts/transcriptHeader.csv',header=T)
colnames(halvedGenes) <- headerHalved$x

```



```{r}
for (i in FCsHalved2$Ensembl_ID[5124:length(FCsHalved2$Ensembl_ID)]){
  transcript2Gene(i)
}




```


```{r}
halvedGenes2 <- read.csv('transcripts/geneTranscriptTable.csv',header=F,sep=',')
colnames(halvedGenes2) <- headerHalved$x

```


The halvedGenes2 has all the genes available for the 75th percentile of genes more than halved in diseased to normal ASCs. The doubledGenes2 and doubledGenes need to be combined.
```{r}
DoubledGenes <- rbind(doubledGenes,doubledGenes2)
HalvedGenes <- halvedGenes2

```

Lets remove those transcripts with 'no ID' from out data.
```{r}
noID_doubled <- grep('no ID',DoubledGenes$transcriptGene)
noID_halved <- grep('no ID',HalvedGenes$transcriptGene)

Doubled <- DoubledGenes[-noID_doubled,]
Halved <- HalvedGenes[-noID_halved,]
```

Lets separate the Ensemble gene IDs from the Gene Symbol IDs for these transcripts.
```{r}
Doubled_ENSG <- grep('ENSG',Doubled$transcriptGene)
Halved_ENSG <- grep('ENSG',Halved$transcriptGene)

Doubled_GeneSymbols <- Doubled[-Doubled_ENSG,]
Halved_GeneSymbols <- Halved[-Halved_ENSG,]

Doubled_Ensembl <- Doubled[Doubled_ENSG,]
Halved_Ensembl <- Halved[Halved_ENSG,]

```

```{r}
write.csv(Doubled_Ensembl,'Doubled_Ensembl.csv',row.names=F)
write.csv(Doubled_GeneSymbols,'Doubled_GeneSymbols.csv',row.names=F)
write.csv(Halved_Ensembl,'Halved_Ensembl.csv',row.names=F)
write.csv(Halved_GeneSymbols,'Halved_GeneSymbols.csv',row.names=F)


```

```{r}
for (i in Doubled_Ensembl$transcriptGene){
  getSummaries2(i,'RNA')
}

for (i in Halved_Ensembl$transcriptGene){
  getSummaries2(i,'transcript')
}

```

```{r}
doubledEnsembl <- read.csv('transcripts/rnasummary.csv',header=F)
halvedEnsembl <- read.csv('transcripts/transcriptsummary.csv',header=F)
rnaHeader <- read.csv('transcripts/geneHeader_summary_rna.csv',header=T)

colnames(doubledEnsembl) <- rnaHeader$x
colnames(halvedEnsembl) <- rnaHeader$x
```


```{r}
for (i in Doubled_GeneSymbols$transcriptGene){
  getSummaries2(i,'RNAs')
}

for (i in Halved_GeneSymbols$transcriptGene){
  getSummaries2(i,'transcripts')
}

```

```{r}
doubledGeneSymbol <- read.csv('transcripts/rnassummary.csv',header=F)
halvedGeneSymbol <- read.csv('transcripts/transcriptssummary.csv',header=F)
colnames(doubledGeneSymbol) <- rnaHeader$x
colnames(halvedGeneSymbol) <- rnaHeader$x
```

Merge by transcripts to the gene summaries.
```{r}
doubleGeneSymbolMerge <- merge(doubledGeneSymbol,Doubled_GeneSymbols,
                     by.x='gene',
                     by.y='transcriptGene')
doubleEnsemblMerge <- merge(doubledEnsembl,Doubled_Ensembl,
                            by.x='gene',
                            by.y='transcriptGene')

DOUBLED75th <- rbind(doubleEnsemblMerge,doubleGeneSymbolMerge)
```

```{r}
halvedGeneSymbolMerge <- merge(halvedGeneSymbol,Halved_GeneSymbols,
                               by.x='gene',
                               by.y='transcriptGene')
halvedEnsemblMerge <- merge(halvedEnsembl,Halved_Ensembl,
                            by.x='gene',
                            by.y='transcriptGene')
HALVED75th <- rbind(halvedGeneSymbolMerge,halvedEnsemblMerge)
```

```{r}
dim(DOUBLED75th);dim(HALVED75th)
```

Now create the data frames of the meta summaries of the transcripts combined with the 75th percentile of Fold changes for those up and down regulated genes by doubled or halved respectively in the disease state as a ratio to the normal state of Adipose Stem Cells.
```{r}
FCsDoubled3 <- merge(DOUBLED75th,FCsDoubled2,
                     by.x='searchedTranscript',by.y='Ensembl_ID')
FCsHalved3 <- merge(HALVED75th,FCsHalved2,
                    by.x='searchedTranscript',by.y='Ensembl_ID')
```

```{r}
FCsDoubled4 <- FCsDoubled3[,-3]
FCsHalved4 <- FCsHalved3[,-3]

FCsDoubled4$gene <- as.character(paste(FCsDoubled4$gene))
FCsDoubled4$EnsemblID <- as.character(paste(FCsDoubled4$EnsemblID))
FCsHalved4$gene <- as.character(paste(FCsHalved4$gene))
FCsHalved4$EnsemblID <- as.character(paste(FCsHalved4$EnsemblID))
```

```{r}
FCsDoubled4$notGene <- grepl('^ENSG',FCsDoubled4$gene)
FCsDoubled4$EnsemblID <- ifelse(FCsDoubled4$notGene==TRUE,
                                FCsDoubled4$gene,
                                FCsDoubled4$EnsemblID)
notGene <- grep('^ENSG',FCsDoubled4$gene)
FCsDoubled4$gene[notGene] <- 'NA'
write.csv(FCsDoubled4,'FCsDoubled75th_summaries.csv',row.names=F)
```

```{r}
FCsHalved4$notGene <- grepl('^ENSG',FCsHalved4$gene)
FCsHalved4$EnsemblID <- ifelse(FCsHalved4$notGene==TRUE,
                                FCsHalved4$gene,
                                FCsHalved4$EnsemblID)
notGene <- grep('^ENSG',FCsHalved4$gene)
FCsHalved4$gene[notGene] <- 'NA'
write.csv(FCsHalved4,'FCsHalved75th_summaries.csv',row.names=F)

```

```{r}
FCsHalved5 <- FCsHalved4[,-c(5:7,32)]
FCsDoubled5 <- FCsDoubled4[,-c(5:7,32)]


write.csv(FCsHalved5,'FCsHalved5_summaryGeneCards.csv',row.names=F)
write.csv(FCsDoubled5,'FCsDoubled5_summaryGeneCards.csv',row.names=F)

```

```{r}
library(tidyr)
```

```{r}
colnames(FCsHalved5
         )
```

```{r}
FCsHalved6 <- gather(FCsHalved5,key='patientID',value='patientValue',5:24)
FCsDoubled6 <- gather(FCsDoubled5,key='patientID',value='patientValue',5:24)
```

```{r}
healthy <- grep('healthy',FCsDoubled6$patientID)
FCsDoubled6$group <- 'diseased ASCs'
FCsDoubled6$group[healthy] <- 'healthy ASCs'

healthy1 <- grep('healthy',FCsHalved6$patientID)
FCsHalved6$group <- 'diseased ASCs'
FCsHalved6$group[healthy1] <- 'healthy ASCs'

```

Tableau Public is unable to process the strings in the transcripts and gene IDs of ENST and ENSG columns, so they will be removed in order to use Tableau Public's graphing propeties.
```{r}


FCsHalved6$searchedTranscript <- gsub('ENST','',FCsHalved6$searchedTranscript)
FCsDoubled6$searchedTranscript <- gsub('ENST','',FCsDoubled6$searchedTranscript)
FCsHalved6$EnsemblID <- gsub('ENSG','',FCsHalved6$EnsemblID)
FCsDoubled6$EnsemblID <- gsub('ENSG','',FCsDoubled6$EnsemblID)

colnames(FCsHalved6)[c(1,3)] <- c('ENST_ensemblTranscript','ENSG_ensemblGene')
colnames(FCsDoubled6)[c(1,3)] <- c('ENST_ensemblTranscript','ENSG_ensemblGene')

```


```{r}
write.csv(FCsHalved6,'FCsHalved75th_summariesTidyGroups.csv',row.names=F)
write.csv(FCsDoubled6,'FCsDoubled75th_summariesTidyGroups.csv',row.names=F)
```

